File: project4/JS/Task2/LinkedList.js
/**
 * @file   LinkedList.js
 * @brief  Implementation of a polymorphic LinkedList data structure in JavaScript.
 * @author Simon Lartey
 * @date   10/29/2025
 *
 * @description
 * This program demonstrates how a LinkedList can be implemented in JavaScript
 * using classes and polymorphism. The LinkedList stores generic data using nodes,
 * similar to how pointers and function handlers work in C. It also includes all
 * major operations: push, pop, append, remove, find, size, clear, and map.
 */

/** Class representing a single Node. */
class Node {
    /**
     * Creates a Node instance.
     * @param {*} data - The data stored in the node.
     */
    constructor(data) {
      this.data = data;
      this.next = null;
    }
  }
  
  /** Class representing a generic LinkedList. */
  class LinkedList {
    constructor() {
      this.head = null;
    }
  
    /** Add a new node to the front of the list. */
    push(data) {
      const newNode = new Node(data);
      newNode.next = this.head;
      this.head = newNode;
    }
  
    /** Remove and return the data from the front node. */
    pop() {
      if (!this.head) return null;
      const value = this.head.data;
      this.head = this.head.next;
      return value;
    }
  
    /** Add a new node to the end of the list. */
    append(data) {
      const newNode = new Node(data);
      if (!this.head) {
        this.head = newNode;
        return;
      }
      let current = this.head;
      while (current.next) current = current.next;
      current.next = newNode;
    }
  
    /**
     * Remove the first node matching target based on a comparison function.
     * @param {*} target - The value to remove.
     * @param {Function} compFunc - Comparison function: returns true if match found.
     */
    remove(target, compFunc) {
      if (!this.head) return null;
  
      if (compFunc(this.head.data, target)) {
        const value = this.head.data;
        this.head = this.head.next;
        return value;
      }
  
      let current = this.head;
      while (current.next) {
        if (compFunc(current.next.data, target)) {
          const value = current.next.data;
          current.next = current.next.next;
          return value;
        }
        current = current.next;
      }
      return null;
    }
  
    /**
     * Find a node matching target using comparison function.
     * @param {*} target - The value to search for.
     * @param {Function} compFunc - Comparison function for equality.
     */
    find(target, compFunc) {
      let current = this.head;
      while (current) {
        if (compFunc(current.data, target)) return current.data;
        current = current.next;
      }
      return null;
    }
  
    /** Return the number of nodes in the list. */
    size() {
      let count = 0;
      let current = this.head;
      while (current) {
        count++;
        current = current.next;
      }
      return count;
    }
  
    /**
     * Remove all nodes, optionally freeing associated resources.
     * @param {Function} freeFunc - Optional cleanup function for node data.
     */
    clear(freeFunc) {
      let current = this.head;
      while (current) {
        if (freeFunc) freeFunc(current.data);
        current = current.next;
      }
      this.head = null;
    }
  
    /**
     * Apply a function to each node’s data.
     * @param {Function} mapFunc - Function to apply to each node.
     */
    map(mapFunc) {
      let current = this.head;
      while (current) {
        mapFunc(current.data);
        current = current.next;
      }
    }
  }
  
  module.exports = { LinkedList };
  

File: Project5/C/linkedlist.c
/**
 * @file   linkedlist.c
 * @brief  Implementation of a generic singly linked list.
 * @author Simon Lartey
 * @date   11/13/2025
 */

#include <stdio.h>
#include <stdlib.h>
#include "linkedlist.h"

/* Create an empty list */
LinkedList *ll_create(void) {
    LinkedList *list = malloc(sizeof(LinkedList));
    if (!list) {
        fprintf(stderr, "Error: Failed to allocate memory for LinkedList.\n");
        return NULL;
    }
    list->head = NULL;
    return list;
}

/* Add node to front */
void ll_push(LinkedList *list, void *data) {
    if (!list) return;

    Node *node = malloc(sizeof(Node));
    if (!node) {
        fprintf(stderr, "Error: Failed to allocate memory for Node.\n");
        return;
    }

    node->data = data;
    node->next = list->head;
    list->head = node;
}

/* Remove node from front */
void *ll_pop(LinkedList *list) {
    if (!list || !list->head) return NULL;

    Node *temp = list->head;
    void *data = temp->data;
    list->head = temp->next;
    free(temp);
    return data;
}

/* Add node to end */
void ll_append(LinkedList *list, void *data) {
    if (!list) return;

    Node *node = malloc(sizeof(Node));
    if (!node) {
        fprintf(stderr, "Error: Failed to allocate memory for Node.\n");
        return;
    }

    node->data = data;
    node->next = NULL;

    if (!list->head) {
        list->head = node;
        return;
    }

    Node *current = list->head;
    while (current->next)
        current = current->next;

    current->next = node;
}

/* Remove first node matching target */
void *ll_remove(LinkedList *list, void *target,
                int (*compfunc)(void *, void *)) {
    if (!list || !list->head) return NULL;

    Node *current = list->head;
    Node *prev = NULL;

    while (current) {
        if (compfunc(current->data, target) == 1) {  // match
            void *data = current->data;

            if (prev)
                prev->next = current->next;
            else
                list->head = current->next;

            free(current);
            return data;
        }
        prev = current;
        current = current->next;
    }
    return NULL;
}

/* Find first node matching target */
void *ll_find(LinkedList *list, void *target,
              int (*compfunc)(void *, void *)) {
    if (!list) return NULL;

    Node *current = list->head;
    while (current) {
        if (compfunc(current->data, target) == 1)
            return current->data;
        current = current->next;
    }
    return NULL;
}

/* Count nodes */
int ll_size(LinkedList *list) {
    int count = 0;
    for (Node *cur = list ? list->head : NULL; cur; cur = cur->next)
        count++;
    return count;
}

/* Clear all nodes */
void ll_clear(LinkedList *list, void (*freefunc)(void *)) {
    if (!list) return;

    Node *current = list->head;
    while (current) {
        Node *next = current->next;
        if (freefunc)
            freefunc(current->data);
        free(current);
        current = next;
    }
    list->head = NULL;
}

/* Apply function to each node’s data */
void ll_map(LinkedList *list, void (*mapfunc)(void *)) {
    if (!list || !mapfunc) return;

    for (Node *cur = list->head; cur; cur = cur->next)
        mapfunc(cur->data);
}
